Allow multiple WebRTC channels per peer

istankovic opened the pull request:
Fixes: #7158.

Commit:
raiden: transport: don't restrict channel creation to lower addresses

Commit:
raiden: transport: rename _RTCPartner to _RTCConnection

Commit:
raiden: transport: move _send_signaling_message to the transport

Commit:
raiden: transport: remove pfs_proxy from WebRTCManager

It's not needed anymore.

Commit:
raiden: transport: propagate signaling send function to _RTCConnection

Commit:
raiden: transport: move candidate handling to _RTCConnection

Commit:
raiden: transport: remove the closed connection callback

Commit:
raiden: transport: move hangup message sending to _RTCConnection

Commit:
raiden: transport: move _handle_sdp_callback to _RTCConnection

Commit:
raiden: transport: create and store call ID when initiating signaling

Also, use the new call ID format with timestamp.

Commit:
raiden: transport: add _RTCConnection.from_offer

Commit:
raiden: transport: create a _RTCConnection when an offer arrives

Commit:
raiden: transport: keep track of connection instances and call IDs

Also create a new connection upon receiving an offer.

Commit:
tests: network: rework the WebRTC test

Commit:
tests: utils: remove unused code

Commit:
raiden: transport: remove an unnecessary function and a debug statement

Commit:
raiden: transport: improve wording in a few places

Commit:
raiden: move WebRTC related types/constants to the WebRTC module

Commit:
raiden: transport: be more robust when processing signaling messages

It only makes sense to have at most two connections per peer, where each
connection was initiatied by a different peer.

netcriptus commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -360,99 +390,79 @@ async def reset(self) -> None:
+        self._signaling_send = signaling_send
+        self._stop_event = stop_event
+        # the addresses for which we're currently trying to initialize WebRTC channels
+        self._web_rtc_channel_inits: Set[Address] = set()
```
Nitpick: If we rename the variable for something else, do you think we could lose the comment? I personally think that something like `self._web_rtc_channels` communicates this idea pretty well.

istankovic replied:
I'm not sure that that would be more appropriate -- note these are not channels, but the addresses of channels we're trying to establish. Most of the time this set is expected to be empty, because all the channels are initialized. It is only _during initialization_ that an address shows up in that set. Hence the `_inits` suffix.

netcriptus commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -360,99 +390,79 @@ async def reset(self) -> None:
+
+    def _handle_message(self, message_data: str, partner_address: Address) -> None:
+        messages: List[ReceivedRaidenMessage] = []
+        for msg in validate_and_parse_message(message_data, partner_address):
```
Doesn't parsing usually include validation? At least Marshmallow and other libraries like that include that step on deserialization as far as I remember.

Would it make sense to have the class `ReceivedRaidenMessage` also do validation and therefore use that to parse and validate instead?

istankovic replied:
> Doesn't parsing usually include validation?

Typically, yes, but it also depends on the kind of validation. E.g. something may be a valid thing, as far as parsing is concerned, but there could be additional validation steps needed (consider, for instance, parsing an integer and checking whether the integer is a square of some other number).

In any case, I just used the existing method here. If there is something to be done, that would be topic for another PR.

netcriptus commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -488,188 +510,179 @@ def _initialize_web_rtc(self, partner_address: Address) -> None:
-        return self._address_to_rtc_partners[partner_address]
+    def _add_connection(self, partner_address: Address, conn: _RTCConnection) -> None:
+        assert (
+            len(self._address_to_connections[partner_address]) < 2
```
Shouldn't this be `<= 2`? The error message leads me to believe that having 2 or fewer connections is ok, but the code will fail for 2 connections.

istankovic replied:
Well, if we already have 2, and we add one, we will have 3, and that is not allowed, as the error message says. So we need to use the `< 2` comparison.

netcriptus reviewed (commented):
Looking good! I just made a few comments there. It would also be great if you could rebase your commits and squash some of them too.

istankovic commented:
> Looking good! I just made a few comments there. It would also be great if you could rebase your commits and squash some of them too.

Thanks for the review, I appreciate it!
I will rebase prior to the merge and I would actually prefer to `merge --no-ff` after the rebase in order to have a clear picture where the work begins and where it ends, since it's a fair amount of changes. As for the squashing, I already did plenty of it and I feel like squashing things further would just make it harder to see what's going on individual commits (it's hard enough as is, for some commits, honestly).

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -228,6 +252,11 @@ async def _set_local_description(self, description: RTCSessionDescription) -> No
+        timestamp = time.time()
+        address1, address2 = sorted((self.node_address, self.partner_address))
+        return f"{to_checksum_address(address1)}|{to_checksum_address(address2)}|{timestamp}"
+
```
would be nice to make sure that `_call_id` is not altered at any point in time. A cached property would probably work.

istankovic replied:
Cached property seems like an overkill for this. I'd rather consider overriding `__setattr__` or something based on the descriptor protocol, but, honestly, I don't think it'd be worth it. The field is private so no outside interference is expected and internally we know how it should be handled.

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -96,7 +113,7 @@ def cancel_all_pending(self) -> None:
-class _RTCPartner(_CoroutineHandler):
+class _RTCConnection(_CoroutineHandler):
     @dataclass
     class SyncEvents:
```
Do we still need sync events? Would be nice to get rid of them (or make them less complex)

istankovic replied:
Unfortunately, we do still need them. However, I just noticed that we could remove `aio_allow_hangup` so I'll do just that.

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -488,188 +510,179 @@ def _initialize_web_rtc(self, partner_address: Address) -> None:
     ) -> None:
-        rtc_partner = self.get_rtc_partner(partner_address)
+        if rtc_message_type == _RTCMessageType.OFFER.value:
+            if self._have_both_connections(partner_address):
```
Couldn't we prevent creating a second channel if there is already one available?

istankovic replied:
Sure, but then we're back to only allowing one channel per peer. Which, if it goes down for any reason, means we would need to establish a new channel again, rather than using the other, presumably still available, channel.

fredo replied:
I think this is not exactly true. If there is a channel alive, while an offer comes in, we can gracefully decline or even ignore the message. This does not imply that only one channel is allowed. Since the process of establishing a channel takes a couple of seconds, two peers could try to create one while the other is not established (channel.readyState == open) yet. In this case we should allow the creation of the two. But if there is a channel available, which means that the other already sees the channel as open (open connection) we can decline the offer. (the offer could come with a bit of latency due to matrix)

Additionally, even if we have two channels and one goes down and we keep using the other one, we end up with having only one channel. Then we are back at the question waht happens if this channel goes down? For this case we should have a retry mechanism in place anyway, so there is no reason to always aim for two channels because if one dies the other eventually will too.

fredo replied:
Maybe it should be noticed that most of the times there is only one user creating a channel. So it is not the default that there are two channels alive.

istankovic replied:
Ok, changed the code to ignore offers if we already have a channel.

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -682,41 +695,34 @@ def _on_ice_gathering_state_change(
 
-    if ice_connection_state in [ICEConnectionState.CLOSED.value, ICEConnectionState.FAILED.value]:
-        asyncio.create_task(rtc_partner.reset())
-        wrap_callback(callback=closed_callback, partner_address=rtc_partner.partner_address)
```
Are you sure it is correct to remove the `closed_callback` here?

istankovic replied:
Well, that callback did two things:
- slept 3s
- tried to establish the WebRTC channel again

My thinking was that was unnecessary. If the channel is closed, the other could still remain available. If no channels are available, then the channel creation will be triggered by health check anyway.

fredo replied:
In my opinion we definitely should have a retry mechanism in place. This would avoid slow payments due to channel failure. Imagine a channel goes down because of a connection issue. Then we should wait for the next healthcheck (which might occur upon the next payment) resulting in a slow payment. 

We should rather retry a fixed amount of times (lets say 3 times) if a channel does not get created, we can assume that the partner is offline and stop retrying, waiting for the next health check. 

This optimistic approach of retrying prevents issues for example when a peer roams to a different network or the network connection breaks (which happens once in a while)

istankovic replied:
Changed to invoke health check that takes care of retrying 3 times.

fredo reviewed (commented):
Very nice clean up of the code! I left some ideas/questions below. 

One thing which comes into my mind, is the retry mechanism already in place if a connection cannot be established?

Commit:
raiden: transport: remove aio_allow_hangup

It's apparently never used.

Commit:
raiden: transport: ignore an incoming offer if we already have a channel

Commit:
raiden: transport: attempt to establish WebRTC channel up to 3 times

Commit:
raiden: transport: invoke health check when an ICE connection closes/fails

Commit:
raiden: transport: no reason for the argument to begin with an underscore

Commit:
raiden: transport: only handle most recent call IDs during signaling

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -682,41 +674,35 @@ def _on_ice_gathering_state_change(
+        _ICEConnectionState.FAILED.value,
+    ]:
+        asyncio.create_task(conn.reset())
+        asyncio.create_task(conn.ice_connection_closed())
```
Why does this need to be its own coroutine? looking at the underlying code it does not have any `await` statement and it only schedules a greenlet which could be done synchronously.

fredo reviewed (changes requested):
I left a comment about `ice_connection_closed`. 

Im also a bit cautious about the whole handling of `ice_connection_state_change`. I dont fully understand if we avoid a situation where a `ice_connection_state == closed/failed` will result in an new establishment process which outdates the current attempt, which triggers a new establishment process which outdates the former, and so on ... we can also discuss in private

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -488,188 +506,162 @@ def _initialize_web_rtc(self, partner_address: Address) -> None:
-
-        return self._address_to_rtc_partners[partner_address]
+    def _add_connection(self, partner_address: Address, conn: _RTCConnection) -> None:
+        assert partner_address not in self._address_to_connection, "must not be there already"
```
I might be mistaken but I think this assert could be hit, because the call in line 510 only schedules a task which will remove the connection eventually but not necessarily directly. since it is not awaited on the loop in _wrapped_initialize_web_rtc could have no context switch and try to add a new connection

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -350,135 +372,131 @@ async def send_message(self, message: str) -> None:
+        if self._stop_event.wait(timeout=_WEB_RTC_CHANNEL_TIMEOUT):
             return
 
         # if room is not None that means we are at least in the second iteration
```
This is legacy comment, could you remove that?

istankovic replied:
Done.

Commit:
raiden: transport: remove stale comment

Commit:
raiden: transport: after ICE conn is closed, only do a health check if we were the initiator

Commit:
raiden: transport: merge reset and close and remove reset_event

This simplfies things further.

Commit:
raiden: transport: make cancel_all_pending private

Commit:
raiden: transport: move close task creation from WebRTCManager to _RTCConnection

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -459,7 +460,10 @@ def _handle_message(self, message_data: str, partner_address: Address) -> None:
+
+        # only do a health check if we were the initiator
+        if conn.call_id.split("|", 1)[0] == to_checksum_address(self.node_address):
+            self.health_check(conn.partner_address)
```
as discussed, I'd prefer a flag check instead of a string comparison of the call_id

fredo replied:
especially when we think about backwards compatibility with the LC, it would be more strict to not rely on the call_id as it has no strict pattern and can be arbitraily chosen. Also nowhere in the code base the checksummed address is used for any logic. The internal representation is always binary

istankovic replied:
Changed to avoid interpreting the call ID.

fredo commented on raiden/network/transport/matrix/rtc/web_rtc.py:
```diff
@@ -584,9 +583,8 @@ def stop(self) -> None:
 
         for partner_address, conn in self._address_to_connection.copy().items():
-            task = self.close_connection(partner_address)
+            self.close_connection(partner_address)
```
Is there a reason why you don't add the `conn.close()` task to all coroutines? will it ever wait on the `close()` coroutine somewhere?

istankovic replied:
Once the task gets to run, it will wait for all the coroutines, in the new `_close` method.

fredo replied:
Yes but I mean where do you wait on this task?

istankovic replied:
This is now handled by the new `wait` method on `_RTCConnection`, that waits for all tasks and greenlets.

Commit:
raiden: transport: move _on_... callbacks to _RTCConnection

Commit:
raiden: transport: remove SyncEvents

We only need aio_allow_candidates.

Commit:
raiden: transport: make set_channel_callbacks private

Also remove the unnecessary condition as the channel always
seems to be set by the time this is called.

Commit:
raiden: transport: introduce channel_open helper on _RTCConnection and use it

Commit:
raiden: transport: make _RTCConnection.channel private

Commit:
raiden: transport: don't rely on call ID to determine the initiator

Commit:
raiden: transport: add an event to control setting remote description

Commit:
raiden: transport: move signaling tasks to _RTCConnection

Commit:
raiden: transport: move send_message and set_candidates tasks to _RTCConnection

Commit:
raiden: transport: make handle_sdp_callback private

Commit:
raiden: transport: unify naming and fix a typo

fredo commented on raiden/tests/unit/network/rtc/test_web_rtc.py:
```diff
@@ -0,0 +1,68 @@
+
+    manager1.health_check(address2)
+    while not manager1.has_ready_channel(address2):
+        gevent.sleep(0.1)
```
around sleeps we usually set timeout context manager, just for faster feedback cycles on the CI. I think the CI default time out is 9 minutes

istankovic replied:
Going to address that in another PR.

Commit:
raiden: transport: use peer connection's connectionState rather than iceConnectionState

It's more reliable.

Commit:
raiden: transport: remove _on_channel_close

It's unneeded because we're anyway watching for connection state
changes. Additionally, close here was never called due to buggy
comparison with our enums.

Commit:
raiden: transport: WebRTCManager does not need to be a _CoroutineHandler anymore

Commit:
raiden: transport: make sure there are no greenlets after stopping

Commit:
raiden: transport: explicitly create and keep track of greenlents in _RTCConnection

This will allow us to remove wrap_callback and to kill all spawned greenlets,
ensuring proper cleanup.

Commit:
raiden: transport: remove create_task_callback and wrap_callback

... and other unneeded things.

Commit:
raiden: transport: introduce a wait method

This waits for both tasks and greenlets.

Commit:
raiden: transport: rename _CoroutineHandler to _TaskHandler

Commit:
raiden: transport: wait on greenlets in _close

Commit:
raiden: transport: add an assert ensure we have a callable, not a future

Commit:
raiden: transport: reset _tasks in wait_for_tasks

Commit:
raiden: transport: improve _RTCConnection.__repr__

fredo reviewed (approved):
LGTM
